espaloma/data/dataset.py

```diff
@@ -0,0 +1,186 @@
+# =============================================================================
+# IMPORTS
+# =============================================================================
+import espaloma as esp
+import abc
+import torch
+
+# =============================================================================
+# MODULE CLASSES
+# =============================================================================
+class Dataset(abc.ABC, torch.utils.data.Dataset):
+    """ The base class of map-style dataset.
+
+    Parameters
+    ----------
+    graphs : list
+        objects in the dataset
+
+    Note
+    ----
+    This also supports iterative-style dataset by deleting `__getitem__`
+    and `__len__` function. 
+
+    Attributes
+    ----------
+    transforms : an iterable of callables that transforms the input.
+
+    """
+    def __init__(self, graphs=None):
+        super(Dataset, self).__init__()
+        self.graphs = graphs
+        self.transforms = None
+
+    def __len__(self):
+        # 0 len if no graphs
+        if self.graphs is None:
+            return 0
+        
+        else:
+            return len(self.graphs)
+
+    def __getitem__(self, idx):
+        if self.graphs is None:
+            raise RuntimeError('Empty molecule dataset.')
+
+        if isinstance(idx, int): # sinlge element
+            if self.transforms is None: # when no transform act like list
+                return self.graphs[idx]
+
+            else:
+                graph = self.graphs[idx]
+
+                # nested transforms
+                for transform in self.transforms:
+                    graph = transform(graph)
+
+                return graph
+
+        elif isinstance(idx, slice): # implement slicing
+            if self.transforms is None:
+                # return a Dataset object rather than list
+                return self.__class__(graphs=self.graphs[idx]) 
+            else:
+                graphs = []
+                for graph in self.graphs:
+
+                    # nested transforms
+                    for transform in transforms:
+                        graph = transform(graph)
+                    graphs.append(graph)
+
+                return self.__class__(graphs=graphs)
+
+    def __iter__(self):
+        if self.transforms is None:
+            return iter(self.graphs)
+
+        else:
+            # TODO:
+            # is this efficient?
+            graphs = iter(self.graphs)
+            for transform in self.transforms:
+                graphs = map(
+                        transform,
+                        graphs)
+
+            return graphs
+
+    def apply(self, fn, in_place=False):
```

Comment:
Docstring would be helpful does `in_place` mean the graphs are modified in place, or that this modifies self.transforms, or both?